module/zfs/dmu_send.c

```diff
@@ -455,6 +468,26 @@ dump_freeobjects(dmu_sendarg_t *dsp, uint64_t firstobj, uint64_t numobjs)
 {
 	struct drr_freeobjects *drrfo = &(dsp->dsa_drr->drr_u.drr_freeobjects);
 
+	/* See comment in dump_free(). */
+	if (!zfs_send_set_freerecords_bit && !dsp->dsa_incremental)
+		return (0);
+
+	if (dsp->dsa_incremental_maxobj > 0) {
+		/*
+		 * drop record, if it is only concerning objects which do not
+		 * exist in base snapshot
+		 */
+		if (dsp->dsa_incremental_maxobj < firstobj)
+			return (0);
+
+		/*
+		 * reduce numobjs according to maximum possible object id of
+		 * base snapshot
```

Comment:
These comments say what is happening, but it isn't much more work to read the code to understand that.

It would be really useful to have a comment explaining *why* we are doing this (especially since the receive code (now) handles longer FREEOBJECTS records just fine).

Reply:
yep, makes sense!